### PR TITLE
Fix Plotter garbage collection

### LINKED ISSUE #2252: Memory Leak Using Plotter and Screenshot
Memory Leak Using Plotter and Screenshot
-----
It seems that `Plotter.screenshot()` produces a memory leak. I am currently running an iterative program implementing the observer pattern, where after each iteration (i.e. when the observer is notified), the observer plots and saves (as a .png) OpenFOAM data.

What I notice in my Activity Monitor is that after each time the observer is called and executed (i.e. PlotVTK.update() is run), my python memory increases by around  ~70MB, which becomes a problem when the observer is called around 300 times, when memory consumption has increased to many many GBs.

I have noticed that when I comment out `p.screenshot(f'plot_{iteration}', window_size=[1000,1000], return_img = False)
`, my memory increase after each observer call is much smaller, i.e. 5MB. Although, when I don't call this observer at all, my memory increase after each observer call in <1MB.

Any help would be appreciated.

**My code**

```py
import Observer
import pyvista as pv

class PlotVTK(Observer):
    
    def update(self, subject) -> None:
        
        # Get current iteration
        iteration = subject.iteration
        
        # Open openfoam file
        reader = pv.OpenFOAMReader(f'test_{iteration}.foam')

        # Obtain data for latest snapshot
        reader.set_active_time_value(reader.time_values[-1])

        # Read data into mesh
        mesh = reader.read()

        # Extract internal mesh data
        internalMesh = mesh["internalMesh"].slice_orthogonal()

        # Create plotting environment
        p = pv.Plotter(off_screen = True)

        # Plot velocity diagram
        p.add_mesh(internalMesh, scalars = 'U')

        # Set viewing angle
        p.view_xy()
        p.camera_position = ((0, 0, 3), (0, 0, 0), (0, 0, 0))
        
        # Take screenshot of image and save in appropriate directory
        p.screenshot(f'plot_{iteration}', window_size=[1000,1000], return_img = False)

        # Clear plot data (i.e. release memory)
        p.clear()
```
-----

**System Information:**
```txt

--------------------------------------------------------------------------------
  Date: Mon Feb 21 16:38:57 2022 GMT

                OS : Darwin
            CPU(s) : 8
           Machine : x86_64
      Architecture : 64bit
               RAM : 16.0 GiB
       Environment : IPython
       File system : apfs
        GPU Vendor : Apple
      GPU Renderer : Apple M1
       GPU Version : 4.1 Metal - 76.3

  Python 3.9.7 (default, Sep 16 2021, 08:50:36)  [Clang 10.0.0 ]

           pyvista : 0.33.2
               vtk : 9.1.0
             numpy : 1.21.2
           imageio : 2.16.0
           appdirs : 1.4.4
            scooby : 0.5.11
        matplotlib : 3.5.0
           IPython : 7.29.0
        ipyvtklink : 0.2.2
             scipy : 1.7.3
              tqdm : 4.62.3
--------------------------------------------------------------------------------

```

### LINKED ISSUE #2577: Plotter memory leak due to not clearing up
@banesullivan I saw this was closed in 2019 however I don't believe the whole problem has been addressed.

Just Running and destroying a pv.Plotter Instance in pyvista 0.32.1 (maybe later but untested) still produces a memory hogging for long running process.

Running the following code: (need to install memory_profiler)

```python

from memory_profiler import memory_usage

import matplotlib.pyplot as plt
import pyvista as pv
import numpy as np
import gc

from time import sleep

gc.enable()


def plot_mem(mem, label=None, show=False):
    if label is None:
        label = 'Usage'
    mem = np.array(mem) - mem[0]
    if show:
        plt.close('all')
        plt.plot(np.arange(len(mem)) * 0.1, mem, label=label)
        plt.ylabel('MB')
        plt.xlabel('time (in s)')  # `memory_usage` logs every 100ms
        plt.legend()
        plt.title('memory usage')
        plt.show()
    print(f'Max: {mem.max():.2f} MB\nLast: {mem[-1]:.2f} MB')


def plot_pyvista():
    # sphere = pv.Sphere()
    for i in range(1000):
        plotter = pv.Plotter()
        plotter.close()

    gc.collect()
    sleep(0.01)


if __name__ == '__main__':
    mem_full = memory_usage((plot_pyvista, [], {}))
    plot_mem(mem_full, label='Full', show=True)
```
We can see that 70 Mib is held over time:

image

Running the following code to try and understand where the memory was being held:

```python
import pyvista as pv

import pandas as pd
from pprint import pprint

from collections import defaultdict
from gc import get_objects
import gc

import sys
before = defaultdict(int)
before_size = defaultdict(int)
after = defaultdict(int)
after_size = defaultdict(int)
for i in get_objects():
    before[type(i)] += 1
    before_size[type(i)] += sys.getsizeof(i)

for i in range(1000):
    plotter = pv.Plotter(off_screen=False)
    plotter.close()

not_collected = gc.collect()


for i in get_objects():
    after[type(i)] += 1
    after_size[type(i)] += sys.getsizeof(i)

df = pd.DataFrame([(str(k), after[k] - before[k], (after_size[k] - before_size[k]))
                  for k in after if after[k] - before[k]], columns=['class', 'count', 'size'])

df = df.groupby('class').sum().reset_index()
pprint(df.sort_values('size').tail(25))
print('\n\n')
pprint(df[['pyvista' in df['class'].astype(str)[i] for i in range(len(df))]])
print('\n\n')
# print(f'{df.size.sum() / 1024:.2f} MB')
print(f"total objects: {df['count'].sum()}")
print(f'not collected objects: {not_collected}')
```

Results:

```
                                                class  count     size
1   <class '_frozen_importlib_external.ExtensionFi...      1       48
11                                   <class 'module'>      1       72
9                         <class 'member_descriptor'>      2      128
2                <class 'builtin_function_or_method'>     17     1224
7                         <class 'getset_descriptor'>     35     2240
25                       <class 'wrapper_descriptor'>     94     6768
22  <class 'vtkmodules.vtkCommonCore.method_descri...    385    27720
3                                      <class 'cell'>    842    33680
16     <class 'pyvista.plotting.renderers.Renderers'>   1000    48000
13        <class 'pyvista.plotting.plotting.Plotter'>   1000    48000
17  <class 'pyvista.plotting.scalar_bars.ScalarBars'>   1000    48000
14  <class 'pyvista.plotting.render_window_interac...   1000    88000
10                                   <class 'method'>   2000   128000
23                                <class 'weakproxy'>   2000   144000
24                                  <class 'weakref'>   2016   145152
6                         <class 'functools.partial'>   2000   160000
12           <class 'pyvista.plotting.camera.Camera'>   2000   176000
15       <class 'pyvista.plotting.renderer.Renderer'>   2000   176000
20                                    <class 'tuple'>   5446   284624
18              <class 'pyvista.themes.DefaultTheme'>   1000   320000
5                                  <class 'function'>   2854   388144
8                                      <class 'list'>  10994   799344
21                                     <class 'type'>    999  1063024
19                                      <class 'set'>   2000  1456000
4                                      <class 'dict'>   8990  3844632



                                                class  count    size
12           <class 'pyvista.plotting.camera.Camera'>   2000  176000
13        <class 'pyvista.plotting.plotting.Plotter'>   1000   48000
14  <class 'pyvista.plotting.render_window_interac...   1000   88000
15       <class 'pyvista.plotting.renderer.Renderer'>   2000  176000
16     <class 'pyvista.plotting.renderers.Renderers'>   1000   48000
17  <class 'pyvista.plotting.scalar_bars.ScalarBars'>   1000   48000
18              <class 'pyvista.themes.DefaultTheme'>   1000  320000



total objects: 49677
not collected objects: 7000
```

It appears a bunch of pyvista objects are not garabge collected I am guessing this is because there are still references between them.

by changing the loop to:

```python
for i in range(1000):
    plotter = pv.Plotter(off_screen=False)
    plotter.close()

    pv.plotting._ALL_PLOTTERS.pop(plotter._id_name)
    plotter.renderers = None
    delattr(plotter, 'renderers')
    plotter._theme = None
    del plotter
```

Results:

```
                                                class  count     size
8                                      <class 'list'>     -6    -8656
0              <class '_frozen_importlib.ModuleSpec'>      1       48
1   <class '_frozen_importlib_external.ExtensionFi...      1       48
16                                <class 'weakproxy'>      1       72
11                                   <class 'module'>      1       72
9                         <class 'member_descriptor'>      2      128
2                <class 'builtin_function_or_method'>     17     1224
7                         <class 'getset_descriptor'>     35     2240
18                       <class 'wrapper_descriptor'>     94     6768
15  <class 'vtkmodules.vtkCommonCore.method_descri...    385    27720
3                                      <class 'cell'>    842    33680
12  <class 'pyvista.plotting.render_window_interac...   1000    88000
10                                   <class 'method'>   2000   128000
17                                  <class 'weakref'>   2016   145152
6                         <class 'functools.partial'>   2000   160000
13                                    <class 'tuple'>   5446   284624
5                                  <class 'function'>   2854   388144
4                                      <class 'dict'>   1990   639904
14                                     <class 'type'>    999  1062728



                                                class  count   size
12  <class 'pyvista.plotting.render_window_interac...   1000  88000



total objects: 19678
not collected objects: 7120
```

The only lingering object appears to be `<class 'pyvista.plotting.render_window_interactor._style_factory.<locals>.CustomStyle'>` 
I managed to trace this down to the init of CustomStyle and in particular the self.AddObserver calls, commenting these lines (and the whole class ...) seems to still provide a working pyvista ....

Running the inital memory plot with all these modifications (commented out CustomStyle) I get:

image

4 Mb so still stuff hanging around but more reasonable.

I am guessing most of the changes can be added to `Plotter.close` just not sure how to handle the CustomStyle

_Originally posted by @WesleyTheGeolien in https://github.com/pyvista/pyvista/issues/482#issuecomment-1104821364_

### PR DESCRIPTION
**Pull request by akaszynski**
This is a less ambitious implementation of #3216 that attempts to tackle just the garbage collection of the `Plotter` class to resolve #2577.

Next step is to try to determine what's causing #2252.

### CODE REVIEW

File: pyvista/plotting/render_window_interactor.py [hunk 1]
```diff
@@ -24,14 +26,17 @@ class RenderWindowInteractor:
...
         if interactor is None:
             interactor = _vtk.vtkRenderWindowInteractor()
+
         self.interactor = interactor
         self.interactor.SetDesiredUpdateRate(desired_update_rate)
         if not light_follow_camera:
             self.interactor.LightFollowCameraOff()
 
-        # Map of observers to events
+        # # Map of observers to events
```
Comment (akaszynski) [inline pyvista/plotting/render_window_interactor.py:39]:
```suggestion
        # Map of observers to events
```

File: pyvista/plotting/render_window_interactor.py [hunk 2]
```diff
@@ -42,10 +47,9 @@ def __init__(self, plotter, desired_update_rate=30, light_follow_camera=True, in
         self._MAX_CLICK_DELAY = 0.8  # seconds
         self._MAX_CLICK_DELTA = 40  # squared => ~6 pixels
 
-        # Set default style
+        # # Set default style
```
Comment (akaszynski) [inline pyvista/plotting/render_window_interactor.py:50]:
```suggestion
        # Set default style
```

File: pyvista/plotting/render_window_interactor.py [hunk 3]
```diff
@@ -24,14 +26,17 @@ class RenderWindowInteractor:
 
     def __init__(self, plotter, desired_update_rate=30, light_follow_camera=True, interactor=None):
         """Initialize."""
+        self.__plotter = weakref.ref(plotter)
+
         if interactor is None:
             interactor = _vtk.vtkRenderWindowInteractor()
+
```
Comment (akaszynski) [inline pyvista/plotting/render_window_interactor.py:33]:
```suggestion
```

**Comment by akaszynski** [comment]
@banesullivan, this is good for review. Five runs, no weird errors.

File: pyvista/plotting/plotting.py [hunk 4]
```diff
@@ -259,7 +258,7 @@ def __init__(
         elif lighting_normalized != 'none':
             raise ValueError(f'Invalid lighting option "{lighting}".')
 
-        # Add self to open plotters
+        # Track all active plotters
```
Comment (akaszynski) [inline pyvista/plotting/plotting.py:261]:
```suggestion
        # Track all active plotters. This has the side effect of ensuring that plotters are not
        # collected until `close()`. See https://github.com/pyvista/pyvista/pull/3216
```

File: pyvista/plotting/render_window_interactor.py [hunk 5]
```diff
@@ -783,6 +791,9 @@ def initialize(self):
 
     def set_render_window(self, ren_win):
         """Set the render window."""
+        # edge_case: accept pyvista.RenderWindow
+        if hasattr(ren_win, '_ren_win'):
+            ren_win = ren_win._ren_win
```
Comment (akaszynski) [inline pyvista/plotting/render_window_interactor.py:796]:
```suggestion
```

File: pyvista/ext/plot_directive.py [hunk 6]
```diff
@@ -373,7 +373,6 @@ def render_figures(
 
             images = []
             figures = pyvista.plotting._ALL_PLOTTERS
```
Comment (akaszynski) [inline pyvista/ext/plot_directive.py:375]:
```suggestion
            figures = pyvista.plotting._ALL_PLOTTERS

```

File: pyvista/plotting/charts.py [hunk 7]
```diff
@@ -1364,6 +1364,7 @@ def show(
         window_size=None,
         notebook=None,
         background='w',
+        **kwargs,
```
Comment (akaszynski) [inline pyvista/plotting/charts.py:1367]:
```suggestion
```

File: pyvista/plotting/charts.py [hunk 8]
```diff
@@ -1404,6 +1405,9 @@ def show(
             ``color='white'``, ``color='w'``, ``color=[1.0, 1.0, 1.0]``, or
             ``color='#FFFFFF'``.  Defaults to ``'w'``.
 
+        **kwargs : dict, optional
+            Optional developer keyword arguments.
+
```
Comment (akaszynski) [inline pyvista/plotting/charts.py:1410]:
```suggestion
```

File: pyvista/plotting/charts.py [hunk 9]
```diff
@@ -1430,6 +1434,7 @@ def show(
         return pl.show(
             screenshot=screenshot,
             full_screen=full_screen,
+            **kwargs,
```
Comment (akaszynski) [inline pyvista/plotting/charts.py:1437]:
```suggestion
```

File: pyvista/plotting/render_window_interactor.py [hunk 10]
```diff
@@ -24,6 +26,8 @@ class RenderWindowInteractor:
 
     def __init__(self, plotter, desired_update_rate=30, light_follow_camera=True, interactor=None):
         """Initialize."""
+        self.__plotter = weakref.ref(plotter)
+
```
Comment (akaszynski) [inline pyvista/plotting/render_window_interactor.py:30]:
```suggestion
```

File: pyvista/plotting/render_window_interactor.py [hunk 11]
```diff
@@ -45,7 +49,6 @@ def __init__(self, plotter, desired_update_rate=30, light_follow_camera=True, in
         # Set default style
         self._style = 'RubberBandPick'
         self._style_class = None
```
Comment (akaszynski) [inline pyvista/plotting/render_window_interactor.py:51]:
```suggestion
        self._style_class = None
        self.__plotter = weakref.ref(plotter)
```

**Comment by larsoner** [comment]
Okay the MNE failures here are reasonable in the sense that we (MNE) modify `_ALL_PLOTTERS` directly, which is a private attribute. IIRC we did this stuff at one point because garbage collection was not always working for us. So we probably need to change how we use `_ALL_PLOTTERS` in concert with this PR.

If the migration is just "don't mess with _ALL_PLOTTERS" then I can open make branch in MNE-Python that does this (or you can) and it can be tested here.

**Comment by akaszynski** [comment]
@larsoner, since this PR effectively fixes the memory leaks with the plotter, I'd add in a backwards compatible try/except or maybe a `pop` so it works for older versions of pyvista.

**Review by larsoner — DISMISSED** [review]
Should come back green with the latest MNE-Python commit -- thanks for working with us on this!

**Comment by akaszynski** [comment]
> Should come back green with the latest MNE-Python commit -- thanks for working with us on this!

More than glad to. I wish we had fixed this a while ago, but late is better than never.

**Pull request merged**